### PR TITLE
read metadata from udta

### PR DESCRIPTION
**Pull request by ririsoft**
Hello,

First I would like to thank you for your work on 'mp4'. I built several implementations of the mp4 standard in Go and Rust for private use that I cannot share, but I would like to join forces to your open initiative and start to contribute.

My main focus for now is metadata parsing : title, release year, description, cover art and more.

I am starting with a _draft pull request_ to gather early feedback from you : is the direction I am pushing the right approach for metadata handling in `mp4-rust` ?

This introduces the 'Metadata' trait to enable access to common video metadata such title, year, cover art and more.
Reading 'title' and 'year' metadata is implemented as a proof of concept.

Using a trait allow's for using deep internal mp4 atom structure and shallow copy (such lossy utf8 conversion) while keeping the access to metadata user friendly at top level `Mp4Reader`.

Once agreed on the read part and `Metadata` API I will implement the write part in the pull request before you merge.

Last : I added a new sample based on [open source ](http://creativecommons.org/licenses/by/3.0/) video [Big Buck Bunny](https://peach.blender.org/about/) with some encoded metadata.

Your comments and request for changes are more than welcome !
Cheers.

### CODE REVIEW

**Comment by alfg** [comment]
Hey @ririsoft, thanks for taking interest in the project. I will take a look soon, thank you!

File: src/types.rs [hunk 1]
```diff
@@ -655,3 +656,63 @@ pub fn creation_time(creation_time: u64) -> u64 {
         creation_time
     }
 }
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+pub enum DataType {
+    Binary = 0x000000,
+    Text = 0x000001,
+    Image = 0x00000D,
+    TempoCpil = 0x000015,
```
Comment (alfg) [inline src/types.rs:665]: Sorry, what type is this?

**Comment by alfg** [comment]
Thanks for the PR. I looked over and I like the approach for the metadata handling. I just had one comment above. Otherwise, feel free to continue. 

Thanks again!

File: src/types.rs (hunk 1 above)
Comment (ririsoft) [reply src/types.rs:665]: Thanks for the review !
Those are the type used to format data in an `ilst` element. They are pretty obvious but `TempoCpil`which I have no idea what it is about. I took it from here: https://xhelmboyx.tripod.com/formats/mp4-layout.txt.
`Binary` is raw vector of u8, `Image`is the same but you know it is an image, `Text`should be valid utf_8 data.

**Comment by ririsoft** [comment]
Thanks !
I will update the PR with more metadata and remove the draft status in a few days.

File: src/types.rs (hunk 1 above)
Comment (alfg) [reply src/types.rs:665]: Yeah, I was referring to the `TempoCpil ` type, but thanks for the clarification! Looks like it's referring to `tmpo/cpil flag = 0x000015`.

Comment (ririsoft) [reply src/types.rs:665]: I have no clue of what it is for and left it for sake of completeness.

**Comment by ririsoft** [comment]
Ok then I rebased onto master and added two more metadata : poster, description.

There are more metadata, but none that I am using currently. Do not hesitate to "@" me in an issue if some users are requesting for a particular metadata not implemented yet.

Also there are other form of metadata, using XML, that can be embedded. I have almost never encountered such. We will see if an issue on that come one day.

Thanks again for your time and review.

**Comment by alfg** [comment]
Sounds good. I'll take a look soon and merge if all is good. Thanks again for your contribution!

**Comment by alfg** [comment]
Hi @ririsoft, I tried a few of the examples using some video files I usually test, but getting a constant `failed to fill whole buffer`.

```
cargo run --example mp4dump tears-of-steel-720p.mp4
    Finished dev [unoptimized + debuginfo] target(s) in 4.02s
     Running `target\debug\examples\mp4dump.exe qt.mp4`
failed to fill whole buffer
```

Same results for other examples as well, including `mp4info`, `mp4sample` and `mp4dump`.

**Comment by ririsoft** [comment]
I can see the issue on Linux not on Mac. I am investigating further and will submit a fix asap.

**Comment by ririsoft** [comment]
Problem identified and should be fixed : a `let start = box_start(reader)?;` was set after reading the box version and thus loosing 4 bytes on the way. Nothing to do with Linux or Mac but rather the video you throw it up. Please double check with your own samples.

**Comment by alfg** [comment]
Thanks for the fix. Will check this out soon.

**Review by alfg — APPROVED** [review]

**Pull request merged**